### ISSUE TITLE
autodiff: params-only backward gives NaN gradients on wgpu/Metal

### ISSUE THREAD

**Issue by laurigates** [tracel-ai/burn#5162]
**Describe the bug**

On Apple Silicon (Metal), training only a small set of adapter parameters (LoRA-style: the whole model is `.no_grad()`, only the low-rank factors are tracked) makes **every gradient come back NaN** on `Autodiff<Wgpu<f16>>` **and** `Autodiff<Wgpu>` (f32), deterministically in a fresh process, while the forward loss is finite. `Autodiff<NdArray>` produces finite gradients for the same graph, and burn-candle's Metal backend (bf16) produces gradients within tolerance of CPU, so the model math itself is sound.

The model is a small DiT-style transformer (2 blocks, width 64: zero-centered RMSNorm, gated-sigmoid GQA attention, pair-interleaved RoPE, 6-way modulation, zero-padding of the token sequence to 256). This is the plain `CubeBackend<WgpuRuntime>` — `burn-fusion` is not in this dependency graph at all.

Four discriminating observations that should narrow the search considerably:

1. **Tracking the model input "fixes" it.** The identical graph with `input.require_grad()` added produces finite gradients for every adapter param. The defect is confined to the *pruned* params-only backward kernel path.
2. **Process state matters.** Running a full (input-tracked) backward first in the same process makes a subsequent params-only backward come back clean — but only for the same float dtype. In a fresh process the params-only backward NaNs every time (3/3 consecutive runs). This smells like a kernel reading uninitialized/out-of-bounds memory (cf. tracel-ai/cubecl#1167), where warm memory pools happen to contain benign data.
3. `CUBECL_AUTOTUNE_LEVEL=0` and `=2` have no effect on the NaN.
4. The burn-store load path is exonerated: loading the same checkpoint on wgpu and reading every param back gives 92/92 tensors byte-identical to a CPU load.

With **real pretrained weights** loaded (instead of random init) there is a second symptom: the *forward* itself NaNs on `Wgpu` f32, while `Wgpu<f16>`'s forward is fine and matches CPU loss to ~1e-3 (loss `cpu=0.8026`, `wgpu-f16=0.8018`, `candle-bf16=0.8047`, `wgpu-f32=NaN`). So the f32 kernels are corrupt in a value-dependent way, not just in the backward.

We tried hard to distill a plain-burn standalone repro and failed — a generic transformer block (several widths from 64 to 512, GQA via `expand`, outer-product pad masks with fully-masked rows, rank-3 broadcast matmul micro-cases) is clean on all backends. The trigger needs something in the DiT graph, so the repro below lives in the (MIT-licensed) project where the model is.

**To Reproduce**

No fixture, no network, no Python needed — random init reproduces it:

```
git clone https://github.com/laurigates/loractl
cd loractl
cargo run --release -p loractl-core --features wgpu --example metal_bisect -- no-load
```

Output on this machine (fresh process, deterministic across runs):

```
ndarray f32    loss=1.158963  loss_finite=true  grads=28  nan_grads=0
wgpu f16       loss=1.029297  loss_finite=true  grads=28  nan_grads=28
wgpu f32       loss=1.029297  loss_finite=true  grads=28  nan_grads=28
```

(the losses differ from CPU because random init differs per backend RNG; the comparison here is finiteness. `examples/grad_compare.rs` is the deterministic-value variant against loaded weights.)

The discriminators are reproducible with the same example:

```
cargo run --release -p loractl-core --features wgpu --example metal_bisect -- adapters
```

which runs single-site adapter subsets × input-tracking on/off — every `igrad` row is finite, and the params-only rows are finite *only* because a same-dtype full backward ran earlier in the process. `-- stages` shows per-stage backward to the inputs is finite everywhere (base model backward is sound). The example's doc comment documents all four modes and observed results.

**Expected behavior**

Params-only backward gradients on wgpu/Metal are finite and match `NdArray` ground truth within dtype tolerance, independent of whether other tensors in the graph are tracked and of prior kernel executions in the process.

**Desktop:**
- Apple M4 Pro, 48 GB unified memory, macOS 26.5.1
- burn 0.21.0 (crates.io), cubecl 0.10.0, wgpu 29.0.4
- rustc 1.96.0

**Additional context**

- Possibly related: tracel-ai/cubecl#1375 (expression-position if/else miscompiles on wgpu/Metal in cubecl 0.10.0 — a silent-wrong-result class in the same backend/version), #3463 (Metal weirdness collection), tracel-ai/cubecl#1120/#1204 (the bulk-write race — fixed in 0.10.0, and our load-path readback confirms loads are clean, so this is a different defect).
- Happy to run instrumented builds / bisect kernel candidates on this hardware — the repro is fast (~5 s a run) and fully deterministic.

**Comment by laurigates** [comment]
Follow-up with a sharper — and in one respect **corrected** — picture. We re-ran the input-tracking discriminator against **loaded weights** (deterministic across backends, so gradients can be compared *numerically*, not just for finiteness). New `workaround` mode in the same example:

```
cargo run --release -p loractl-core --features wgpu --example metal_bisect -- workaround <bundle>
```

Truth table on the same graph and weights:

| arm | forward loss | adapter B-grads |
|---|---|---|
| ndarray f32 | 0.802559 | ground truth (14 sites) |
| wgpu f32, params-only | **NaN (the forward!)** | — |
| wgpu f32 + input tracked | **0.802559 — bit-identical to CPU** | **ratio 1.000 on all 14 sites** |
| wgpu f16, params-only | 0.8018 (≈ correct) | all NaN |
| wgpu f16 + input tracked | **0.7771 (wrong)** | **exactly 0.0** — unchanged at loss scale 64 vs 16384 |

Three takeaways:

1. **The correction:** "tracking the input heals it" (from the random-init runs in the original report) is *value-dependent*, not structural. With loaded weights, tracking makes wgpu-**f32** bit-exact — but wgpu-**f16** flips to a different failure: a wrong-but-finite forward and exactly-zero gradients. Zero grads + shifted loss + loss-scale invariance rules out f16 range saturation; values are being dropped, not overflowed.
2. Tracking one extra tensor flips wgpu-f32 between a NaN *forward* and bit-exact correctness on identical inputs. Since tracking only changes where autodiff retains intermediates — i.e. where the lazy execution stream gets cut into kernels — the defect looks like a specific kernel/boundary configuration that only materializes for some graph cuts, consistent with the process-warm-state healing in the original report.
3. The exactly-zero f16 gradients match the corruption signature we originally chased on the real 12B model (adapter `B` stayed identically 0 through optimizer steps), so the tiny repro is faithful to the production failure.

Repro at loractl `main` (`crates/loractl-core/examples/metal_bisect.rs`, modes documented in the header; the `<bundle>` fixture for `workaround` is generated offline by `just krea2-reference`, needs `uv`+torch — the fixture-free `no-load`/`adapters`/`stages` modes need nothing).

**Comment by laurigates** [comment]
Cross-platform data points from a Linux/RTX 4090 box: **a Metal-compiler-only explanation is ruled out — the params-only backward also breaks on wgpu-Vulkan, and the f16 signatures reproduce on CUDA. CUDA f32 is fully clean.** (f32 behavior diverges per backend, so more than one defect may be in play; details below.)

Setup: identical to the original report — a `.no_grad()` model with the LoRA adapter A/B params as the only tracked tensors, loaded weights, deterministic identical inputs, one forward+backward per backend, per-site max-|grad| compared against ndarray f32 (a "site" = one adapter injection point; "ratio 1.00" means the GPU/CPU ratio of that per-site max-|grad|, matching to ~4 significant digits at every site). burn 0.21 / cubecl 0.10. All arms run from **one binary built with both features** (`--features cuda,wgpu`; nvcc 13.0 at build time, driver 580.126.18); the CUDA result was additionally re-confirmed from a build with **only** the `cuda` feature, ruling out a dual-linkage interaction.

| arm (params-only tracked, loaded weights) | forward loss | tracked-param grads |
|---|---|---|
| ndarray f32 (reference) | 0.802559 | reference |
| **CUDA f32** | 0.802561 | **ratio 1.00 vs CPU at every site** ✅ |
| CUDA f16 | **0.777105** | exactly 0.0 at most sites (a few denormals ~1e-42) |
| wgpu-Vulkan f32 | **NaN** | readback garbage (below) |
| wgpu-Vulkan f16 | 0.8018 (≈ correct) | readback garbage (below) |
| wgpu-Vulkan f32 + one input tensor additionally tracked | 0.8026, matches CPU | ratio 1.000 every site — the same workaround as on Metal |
| wgpu-Vulkan f16 + input tracked | — | exactly 0.0, same as Metal |

**"Readback garbage":** the diagnostic computes `grad.abs().max()` on-device, then reads the scalar back to the host — and **both Vulkan arms return the same impossible value, −3.998**, for that op chain. The impossibility is itself the finding: on Vulkan-with-loaded-weights, either the reduction kernels or the device→host readback of the params-only backward's outputs is corrupt (the text can't distinguish which). The identical instrumentation in the same binary reads back correct values on ndarray, CUDA f32, and the input-tracked Vulkan arm, which argues against the diagnostic itself being racy.

**How the f16 values map onto the Metal table:** the original report has two f16 rows — params-only (forward 0.8018 ≈ correct, grads all-NaN) and input-tracked (forward **0.7771 wrong**, grads exactly 0.0). Here, Vulkan f16 params-only reproduces the *params-only* row's forward (0.8018), while **CUDA f16 params-only produces 0.7771 — the exact wrong value Metal produces in its input-tracked configuration**. Two backends independently arriving at the same wrong forward value looks like a deterministic wrong computation in a shared path rather than precision noise, but I have no hypothesis for why the two backends land on *different* Metal rows' values — reported as observed; these may be convergent distinct defects.

The original report's discriminators hold on Vulkan: load path clean (92/92 tensors byte-exact after a `SafetensorsStore` load), backward-to-inputs finite at every stage, only the params-only pruned backward breaks, and tracking one extra input tensor heals the f32 arm completely. One divergence worth having as a discriminator: the **random-init variant (no file load) is fully clean on Vulkan** — both f32 and f16, 0/28 NaN grads — whereas on Metal it breaks fixture-free (re-confirmed today on the same Apple-Silicon host as the original report: 28/28 NaN grads on both f16 and f32). On Vulkan the corruption requires loaded weights; on Metal it does not.

Evidence for "CUDA f32 fully clean" beyond the table row: a 120-step LoRA training smoke (f32 + gradient checkpointing) passes — `test cuda_training_smoke ... ok` (2.96s, finite monotonically-decreasing loss, adapter written, reloaded on CUDA with finite logits) — and an end-to-end CLI training run on the same backend completes and writes a loadable adapter.

On the repro: as noted in the original report, six attempts to distill this to a plain-burn standalone case (generic transformer block, expand-GQA, pad-row masks, broadcast-matmul micro-repros) all pass — the trigger seems to need the full model graph. So the runnable repro remains the original one, now with CUDA arms added (laurigates/loractl#94): `cargo run --release -p loractl-core --features cuda,wgpu --example grad_compare -- crates/loractl-core/tests/fixtures/tiny-krea2` (fixture checked in; the same repo's `metal_bisect` example runs the verify-load / stages / no-load / workaround discriminators).

**Comment by laurigates** [comment]
A third data point that narrows the f16 case further: **quantizing the frozen base to int8 — so the base linears run through a hand-written custom `Backward` op instead of the stock `matmul` — does not change the f16 outcome at all.** The exactly-zero f16 adapter gradients are independent of the base op.

Motivation: one hypothesis for the "exactly 0.0 f16 grads" signature is that it lives in how autodiff checkpoints/prunes the frozen base matmul's backward. The QLoRA path is a clean way to test that, because it replaces the base op with a *different* backward implementation:

- The base `Linear`s are quantized to weight-only int8 and carried through the autodiff graph as constants via a custom `impl Backward` (`x · dequant(wq)ᵀ` forward; `grad_x = grad_out · dequant(wq)` backward). That op **tracks the activation `x` as its parent and never checkpoints the weight** — structurally the same "an input is tracked" shape that heals the *f32* arm in the tables above.
- The LoRA `A`/`B` adapters remain plain f32 tensors, exactly as before.

Same setup as my previous comment (`.no_grad()` model, adapters the only tracked params, loaded weights, deterministic identical inputs, one binary `--features cuda,wgpu`, stock cubecl 0.10 — no fork), run for both a **plain** base and a **quantized** base:

| arm | PLAIN base (mean adapter-B grad ratio vs CPU) | QUANTIZED base / custom op (same) |
|---|---|---|
| ndarray f32 (reference) | 1.00 | 1.00 |
| CUDA f32 | 1.00 ✅ | **1.00 ✅** |
| CUDA f16 | **0.00** | **0.00 (identical)** |
| wgpu-Vulkan f16 / f32 | garbage / NaN | garbage / NaN |

So on CUDA f16, the adapter-B gradients come back exactly zero **whether the base is a stock `matmul` or the activation-tracking custom op** — the loss is the same wrong 0.777105 in both. Two consequences, reported as observed:

1. The f16 zero-grad defect is **not** in the frozen base's backward/checkpointing — it survives replacing that op wholesale. The `A`/`B` adapters ride the generic f16 autodiff path regardless of the base, so the corruption is in the adapter gradient path itself (or a shared f16 autodiff kernel it uses), upstream of any single op one can hand-write.
2. The "track an input to heal it" workaround that fixes the *f32* arm does **not** generalize to f16 by putting the tracked-activation shape on the base op — the custom op has exactly that shape and the f16 grads are still zero. Consistent with the earlier finding that f16 input-tracking flips to a *wrong-but-finite* forward with zero grads rather than healing.

Bonus cross-check: the plain-base arms in this run reproduce the CUDA table from my previous comment exactly (CUDA f32 clean at 1.00; CUDA f16 → 0.777105 forward + zero grads), and CUDA f32 stays numerically clean under quantization too (grad ratio 1.00) — so it's specifically the f16 route this blocks, not int8 QLoRA per se.

Repro (checked-in fixture, no network/Python; the same tiny DiT graph as the other examples):

```
cargo run --release -p loractl-core --features cuda,wgpu \
  --example quant_grad_compare -- crates/loractl-core/tests/fixtures/tiny-krea2
```

(runs each backend arm twice — plain base, then quantized base — and prints the two tables above; drop `cuda` from the features for the wgpu-only arms.)
